### ISSUE TITLE
Armeria server does not handle the `Expect: 100-continue` header correctly.

### ISSUE THREAD

**Issue by trustin** [line/armeria#1254]
`Http1ObjectDecoder` and `Http2ObjectDecoder` does not send `100 Continue` informational header when a client sends a request with the `Expect: 100-continue` header, which is obviously a violation of the HTTP specification, unless we want to reject the request.

Reported by @jwills

**Comment by trustin** [comment]
One problem with implementing this at the decoder level is that a `Service` will not be given a way to reject the request before a client sends the body because the decoder will always just respond with `100 Continue`. This is not ideal for a service with large request payloads such as file upload service.

**Comment by trustin** [comment]
Alternatively, we may be able to allow a `Service` to choose to handle an `Expect` header by handling the `Expect` header in `HttpServerHandler`:

- `HttpServerHandler` queries a service is it desires to handle the `Expect` header by itself.
- If so, pass the request as it is and do not perform any additional processing.
- If not, remove the `Expect` header from the request headers so that the `Service` does not see the `Expect` header and send `100 Continue` response automatically.

One tiny remaining problem is handling an HTTP/1.0 request. According to the RFC, we have to ignore the `Expect` header if the client is an HTTP/1.0 client. However, `HttpServerHandler` does not have such information; it just knows whether it's HTTP/1.x or HTTP/2. Therefore, we may pass the exact HTTP version to `HttpServerHandler`.

**Comment by trustin** [comment]
How do we let a user tell us his or her `Service` will handle an `Expect` header by itself? I think we can introduce a tag interface like we did `TransientService`, such as, `ExpectAwareHttpService`? Naming suggestion appreciated.

**Comment by trustin** [comment]
After thinking a little bit more, I realize it's harder than initially thought. For example, let's say a user decorates his or her `Service` with various decorators which may want to reject the request, such as `HttpAuthService`. In this setup, some decorators may need to check the `Expect` header but others may not want to do so.

**Comment by trustin** [comment]
Again, this is also harder than expected, because the `Service` implementation would become way more complicated due to different way of building a response, e.g.

```java
if (expectHeaderExists) {
    HttpResponseWriter res = HttpResponse.streaming();
    res.write(HttpHeaders.of(100));
    // ... handle the request and write to 'res' ...
    return res;
} else {
    // ... handle the request ...
    HttpResponse res = HttpResponse.of(200, ...);
    return res;
}
```

**Comment by jwills** [comment]
Is this something I can simply handle myself in my own server impl?

**Comment by minwoox** [comment]
Closed by #1255
